contrib/src/main/java/org/apache/apex/malhar/contrib/kudu/AbstractKuduOutputOperator.java

```diff
@@ -0,0 +1,666 @@
+ * <li>Define the payload class</li>
+ * </ol>
+ * </p>
+ * <p>
+ * Note that the tuple that is getting processed is not the POJO class. The tuple that is getting processed is the
+ * {@link KuduExecutionContext} instance. This is because the operator supports mutation types as a higher level
+ * construct than simply writing a POJO into a Kudu table row.
+ * </p>
+ * <p>
+ * Supported mutations are:
+ * <ol>
+ *      <li>INSERT</li>
+ *      <li>UPDATE</li>
+ *      <li>UPSERT</li>
+ *      <li>DELETE</li>
+ * </ol>
+ * </p>
+ * <p>
+ * Please note that the Update mutation allows to change a subset of columns only even though there might be columns
+ * that were defined to be non-nullable. This is because the original mutation of type insert would have written the
+ * mandatory columns. In such scenarios, the method setDoNotWriteColumns() in {@link KuduExecutionContext} can be
+ * used to specify only those columns that need an update. This ways a read and Update pattern can be merged to a
+ * simple update pattern thus avoiding a read if required.</p>
+ * */
+@InterfaceStability.Evolving
+public abstract class AbstractKuduOutputOperator extends BaseOperator
+    implements Operator.ActivationListener<Context.OperatorContext>,Operator.CheckpointNotificationListener
+{
+
+  private ApexKuduConnection apexKuduConnection;
+
+  private transient  KuduTable kuduTable;
+
+  private transient  KuduSession kuduSession;
+
+  private transient KuduClient kuduClientHandle;
+
+  private transient Map<String,ColumnSchema> allColumnDefs;
+
+  private transient Map<String,Object> kuduColumnBasedGetters;
+
+  private Set<String> primaryKeyColumnNames;
+
+  private static final transient Logger LOG = LoggerFactory.getLogger(AbstractKuduOutputOperator.class);
+
+  private WindowDataManager windowDataManager;
+
+  private transient long currentWindowId;
+
+  private transient boolean isInSafeMode;
+
+  private transient boolean isInReconcilingMode;
+
+  private transient long reconcilingWindowId;
+
+  @AutoMetric
+  transient long numInsertsInThisWindow;
+
+  @AutoMetric
+  transient long numUpsertsInThisWindow;
+
+  @AutoMetric
+  transient long numDeletedInThisWindow;
+
+  @AutoMetric
+  transient long numUpdatesInThisWindow;
+
+  @AutoMetric
+  transient long opsErrorsInThisWindow;
+
+  @AutoMetric
+  transient long bytesWrittenInThisWindow;
+
+  @AutoMetric
+  transient long rpcErrorsInThisWindow;
+
+  @AutoMetric
+  transient long writeOpsInThisWindow;
+
+  @AutoMetric
+  transient long writeRPCsInThisWindow;
+
+  @AutoMetric
+  long totalOpsErrors = 0;
+
+  @AutoMetric
+  long totalBytesWritten = 0;
+
+  @AutoMetric
+  long totalRpcErrors = 0;
+
+  @AutoMetric
+  long totalWriteOps = 0;
+
+  @AutoMetric
+  long totalWriteRPCs = 0;
+
+  @AutoMetric
+  long totalInsertsSinceStart;
+
+  @AutoMetric
+  long totalUpsertsSinceStart;
+
+  @AutoMetric
+  long totalDeletesSinceStart;
+
+  @AutoMetric
+  long totalUpdatesSinceStart;
+
+  @InputPortFieldAnnotation
```

Comment:
I can do that .. A quick question though .. Do we use the annotation for any special processing inside the apex core ? i.e. get all annotated fields that have this annotation and do any extra processing ? What is the annotation used for if it is something that can be ignored in the first place ?